<a href="https://colab.research.google.com/github/antonychackotc/project-1/blob/main/Retail_Order_Data_Analysis_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
!kaggle datasets download ankitbansal06/retail-orders -f orders.csv


Dataset URL: https://www.kaggle.com/datasets/ankitbansal06/retail-orders
License(s): CC0-1.0
  0% 0.00/200k [00:00<?, ?B/s]
100% 200k/200k [00:00<00:00, 43.0MB/s]


In [2]:
import zipfile
with zipfile.ZipFile("/content/orders.csv.zip","r") as zip_ref:
    zip_ref.extractall("/content/")

In [3]:
import pandas as pd
df=pd.read_csv('/content/orders.csv')
df.head(5)

Order Id  Order Date       Ship Mode    Segment        Country  \
0         1  2023-03-01    Second Class   Consumer  United States   
1         2  2023-08-15    Second Class   Consumer  United States   
2         3  2023-01-10    Second Class  Corporate  United States   
3         4  2022-06-18  Standard Class   Consumer  United States   
4         5  2022-07-13  Standard Class   Consumer  United States   

              City       State  Postal Code Region         Category  \
0        Henderson    Kentucky        42420  South        Furniture   
1        Henderson    Kentucky        42420  South        Furniture   
2      Los Angeles  California        90036   West  Office Supplies   
3  Fort Lauderdale     Florida        33311  South        Furniture   
4  Fort Lauderdale     Florida        33311  South  Office Supplies   

  Sub Category       Product Id  cost price  List Price  Quantity  \
0    Bookcases  FUR-BO-10001798         240         260         2   
1       Chairs  FUR-CH-10000454         600         730         3   
2       Labels  OFF-LA-10000240          10          10         2   
3       Tables  FUR-TA-10000577         780         960         5   
4      Storage  OFF-ST-10000760          20          20         2   

   Discount Percent  
0                 2  
1                 3  
2                 5  
3                 2  
4                 5

# **Handling Missing Values**

In [4]:
df.isnull().sum().sum()

1

In [5]:
df.fillna(0,inplace=True)

In [6]:
df.isnull().sum().sum()

0

# **Renaming Columns:**

In [7]:
df.rename({'Order Id':'Order_Id','Order Date':'Order_Date','Ship Mode':'Ship_Mode','Postal Code':'Postal_Code','Sub Category':'Sub_Category','Product Id':'Product_Id','cost price':'cost_price','List Price':'List_Price','Discount Percent':'Discount_Percent'},axis=1,inplace=True)

In [8]:
df.head(3)

Order_Id  Order_Date     Ship_Mode    Segment        Country         City  \
0         1  2023-03-01  Second Class   Consumer  United States    Henderson   
1         2  2023-08-15  Second Class   Consumer  United States    Henderson   
2         3  2023-01-10  Second Class  Corporate  United States  Los Angeles   

        State  Postal_Code Region         Category Sub_Category  \
0    Kentucky        42420  South        Furniture    Bookcases   
1    Kentucky        42420  South        Furniture       Chairs   
2  California        90036   West  Office Supplies       Labels   

        Product_Id  cost_price  List_Price  Quantity  Discount_Percent  
0  FUR-BO-10001798         240         260         2                 2  
1  FUR-CH-10000454         600         730         3                 3  
2  OFF-LA-10000240          10          10         2                 5

# **Trimming Spaces**

In [9]:
cols = ['Order_Id','Order_Date','Ship_Mode','Segment','Country','City','State','Postal_Code','Region','Category','Sub_Category','Product_Id','cost_price','List_Price','Quantity','Discount_Percent']

for col in cols:
    df[col] = df[col].astype(str).str.strip()
df = df[cols]
df.head(3)

Order_Id  Order_Date     Ship_Mode    Segment        Country         City  \
0        1  2023-03-01  Second Class   Consumer  United States    Henderson   
1        2  2023-08-15  Second Class   Consumer  United States    Henderson   
2        3  2023-01-10  Second Class  Corporate  United States  Los Angeles   

        State Postal_Code Region         Category Sub_Category  \
0    Kentucky       42420  South        Furniture    Bookcases   
1    Kentucky       42420  South        Furniture       Chairs   
2  California       90036   West  Office Supplies       Labels   

        Product_Id cost_price List_Price Quantity Discount_Percent  
0  FUR-BO-10001798        240        260        2                2  
1  FUR-CH-10000454        600        730        3                3  
2  OFF-LA-10000240         10         10        2                5

# **CALCULATING DISCOUNTS-**

# **Selling Price = Listed Price [(100−discount%)/100]**

In [10]:
df['Discount_Percent'] = pd.to_numeric(df['Discount_Percent'], errors='coerce')
df['rough_column'] = (100 - df['Discount_Percent'])

In [11]:
df['rough_column_1']=df['rough_column']/100

In [12]:
df['List_Price'] = pd.to_numeric(df['List_Price'], errors='coerce')

df['Selling_Price']=df['List_Price']*df['rough_column_1']

In [13]:
df.drop(['rough_column','rough_column_1'],axis=1,inplace=True)


In [14]:
df.head(3)

Order_Id  Order_Date     Ship_Mode    Segment        Country         City  \
0        1  2023-03-01  Second Class   Consumer  United States    Henderson   
1        2  2023-08-15  Second Class   Consumer  United States    Henderson   
2        3  2023-01-10  Second Class  Corporate  United States  Los Angeles   

        State Postal_Code Region         Category Sub_Category  \
0    Kentucky       42420  South        Furniture    Bookcases   
1    Kentucky       42420  South        Furniture       Chairs   
2  California       90036   West  Office Supplies       Labels   

        Product_Id cost_price  List_Price Quantity  Discount_Percent  \
0  FUR-BO-10001798        240         260        2                 2   
1  FUR-CH-10000454        600         730        3                 3   
2  OFF-LA-10000240         10          10        2                 5   

   Selling_Price  
0          254.8  
1          708.1  
2            9.5

# **Discount = List Price - Selling Price**

In [15]:
df['Discount']=df['List_Price']-df['Selling_Price']

In [16]:
df.head(3)

Order_Id  Order_Date     Ship_Mode    Segment        Country         City  \
0        1  2023-03-01  Second Class   Consumer  United States    Henderson   
1        2  2023-08-15  Second Class   Consumer  United States    Henderson   
2        3  2023-01-10  Second Class  Corporate  United States  Los Angeles   

        State Postal_Code Region         Category Sub_Category  \
0    Kentucky       42420  South        Furniture    Bookcases   
1    Kentucky       42420  South        Furniture       Chairs   
2  California       90036   West  Office Supplies       Labels   

        Product_Id cost_price  List_Price Quantity  Discount_Percent  \
0  FUR-BO-10001798        240         260        2                 2   
1  FUR-CH-10000454        600         730        3                 3   
2  OFF-LA-10000240         10          10        2                 5   

   Selling_Price  Discount  
0          254.8       5.2  
1          708.1      21.9  
2            9.5       0.5

# **Profit = Selling Price - Cost Price**

In [17]:
df['cost_price'] = pd.to_numeric(df['cost_price'], errors='coerce')

df['Profit']=df['Selling_Price']-df['cost_price']
df.head(3)

Order_Id  Order_Date     Ship_Mode    Segment        Country         City  \
0        1  2023-03-01  Second Class   Consumer  United States    Henderson   
1        2  2023-08-15  Second Class   Consumer  United States    Henderson   
2        3  2023-01-10  Second Class  Corporate  United States  Los Angeles   

        State Postal_Code Region         Category Sub_Category  \
0    Kentucky       42420  South        Furniture    Bookcases   
1    Kentucky       42420  South        Furniture       Chairs   
2  California       90036   West  Office Supplies       Labels   

        Product_Id  cost_price  List_Price Quantity  Discount_Percent  \
0  FUR-BO-10001798         240         260        2                 2   
1  FUR-CH-10000454         600         730        3                 3   
2  OFF-LA-10000240          10          10        2                 5   

   Selling_Price  Discount  Profit  
0          254.8       5.2    14.8  
1          708.1      21.9   108.1  
2            9.5       0.5    -0.5

In [18]:
df.to_csv('cleaned-orders.csv')

year, month and day extract from order_date using of to_datetime

In [19]:
df['Order_Date']=pd.to_datetime(df['Order_Date'])
df['year']=df['Order_Date'].dt.year
df['month']=df['Order_Date'].dt.month
df['day']=df['Order_Date'].dt.day
df.head(3)

Order_Id Order_Date     Ship_Mode    Segment        Country         City  \
0        1 2023-03-01  Second Class   Consumer  United States    Henderson   
1        2 2023-08-15  Second Class   Consumer  United States    Henderson   
2        3 2023-01-10  Second Class  Corporate  United States  Los Angeles   

        State Postal_Code Region         Category  ... cost_price List_Price  \
0    Kentucky       42420  South        Furniture  ...        240        260   
1    Kentucky       42420  South        Furniture  ...        600        730   
2  California       90036   West  Office Supplies  ...         10         10   

   Quantity  Discount_Percent Selling_Price  Discount  Profit  year  month  \
0         2                 2         254.8       5.2    14.8  2023      3   
1         3                 3         708.1      21.9   108.1  2023      8   
2         2                 5           9.5       0.5    -0.5  2023      1   

   day  
0    1  
1   15  
2   10  

[3 rows x 22 columns]

Month name extract from month value using maping

In [20]:
dw_maping={

1:'January',
2:'February',
3:'March',
4:'April',
5:'May',
6:'June',
7:'July',
8:'August',
9:'September',
10:'October',
11:'November',
12:'December'
}
df['Month_name']=df['Order_Date'].dt.month.map(dw_maping)

In [21]:
df.head(3)

Order_Id Order_Date     Ship_Mode    Segment        Country         City  \
0        1 2023-03-01  Second Class   Consumer  United States    Henderson   
1        2 2023-08-15  Second Class   Consumer  United States    Henderson   
2        3 2023-01-10  Second Class  Corporate  United States  Los Angeles   

        State Postal_Code Region         Category  ... List_Price Quantity  \
0    Kentucky       42420  South        Furniture  ...        260        2   
1    Kentucky       42420  South        Furniture  ...        730        3   
2  California       90036   West  Office Supplies  ...         10        2   

   Discount_Percent  Selling_Price Discount  Profit  year  month  day  \
0                 2          254.8      5.2    14.8  2023      3    1   
1                 3          708.1     21.9   108.1  2023      8   15   
2                 5            9.5      0.5    -0.5  2023      1   10   

   Month_name  
0       March  
1      August  
2     January  

[3 rows x 23 columns]

In [22]:
df.to_csv('cleaned-orders-1.csv')

# **process of localhost vscode or jupiter**

pip install mysql-connector-python

import mysql.connector

mydb=mysql.connector.connect

(

host='localhost',

user='root',

password='pass@123',

auth_plugin='mysql_native_password'

)
***************************************************
mycursor=mydb.cursor()

mycursor.execute('create database clearorder1')

mycursor.execute('create table orders')

mycursor.execute('show database')

for x in mycursor:
 print(x)

mycursor.execute('use clearorder1')

*********************************************************************************

mycusor=mydb.cursor()

mycursor.execute('select * from cleardata1')

myresult=mycursor.fetchall()

for x in myresult:
 print(x)

In [23]:
# SQL

# **SQL**

In [24]:
import sqlite3
import pandas as pd

# File name of the uploaded CSV
csv_filename = "/content/cleaned-orders-1.csv"  # Replace this with the uploaded file name

# Load CSV into a pandas DataFrame
df = pd.read_csv(csv_filename)

# Connect to SQLite database (or create a new one)
conn = sqlite3.connect("example.db")
cursor = conn.cursor()

# Write DataFrame to SQLite table
table_name = "cleanedorders1"  # Specify your table name
df.to_sql(table_name, conn, if_exists="replace", index=False)

print(f"Table '{table_name}' created in SQLite database.")


Table 'cleanedorders1' created in SQLite database.


# **Just count of all datas**

In [25]:
query = f"SELECT COUNT(*) FROM {table_name}"
result = pd.read_sql_query(query, conn)

# Display the results
result

COUNT(*)
0      9994

# **Average selling price**

In [26]:
query = f"SELECT product_id,avg(selling_price) FROM cleanedorders1;"
result = pd.read_sql_query(query, conn)

# Display the results
result

Product_Id  avg(selling_price)
0  FUR-BO-10001798          221.718901

# **#Maximum Selling Price**

In [27]:
query = f"SELECT order_id,product_id,max(selling_price) FROM cleanedorders1;"
result = pd.read_sql_query(query, conn)

# Display the results
result

Order_Id       Product_Id  max(selling_price)
0      2698  TEC-MA-10002412             21734.4

#Minimum Selling Price


In [28]:
query = f"SELECT order_id,product_id,min(selling_price) FROM cleanedorders1;"
result = pd.read_sql_query(query, conn)

# Display the results
result

Order_Id       Product_Id  min(selling_price)
0        16  OFF-BI-10000756                 0.0

**Top-Selling Products: Identify the products that generate the highest revenue based on sale prices.**

In [29]:
query = f"SELECT row_number() over(order by product_id) as row,product_id, selling_price from cleanedorders1 where selling_price order by selling_price Desc;"
result = pd.read_sql_query(query, conn)

# Display the results
result

row       Product_Id  Selling_Price
0     8546  TEC-MA-10002412        21734.4
1     8484  TEC-CO-10004722        16975.0
2     8485  TEC-CO-10004722        13440.0
3     8481  TEC-CO-10004722        10976.0
4     8482  TEC-CO-10004722         9975.0
...    ...              ...            ...
9482  7945  TEC-AC-10002018            9.5
9483  8645  TEC-PH-10000307            9.5
9484  8656  TEC-PH-10000376            9.5
9485  8984  TEC-PH-10002185            9.5
9486  8987  TEC-PH-10002185            9.5

[9487 rows x 3 columns]

# **Monthly Sales Analysis**

In [30]:
# select order_date from cleanedorders1 where year order by year asc;

# SELECT count(*) order_date FROM cleanedorders1 WHERE month_name='January' and year='2022';
# SELECT order_date FROM cleanedorders1 WHERE month_name='January' and year='2022';

In [31]:
query = f"SELECT count(*) order_date FROM cleanedorders1 WHERE month_name='January' and year='2022';"
result = pd.read_sql_query(query, conn)

# Display the results
result

order_date
0         439

In [32]:
# Maximum Sales

# SELECT month_name,year,day,product_id,order_id,max(Selling_Price) FROM cleanedorders1;

# Minimum Sales

# SELECT month_name,year,day,product_id,order_id,min(Selling_Price) FROM cleanedorders1;

# Average Sales

# SELECT month_name,year,day,product_id,order_id,max(Selling_Price) FROM cleanedorders1;

In [33]:
query = f"SELECT month_name,year,max(Selling_Price) as sales FROM cleanedorders1 group by month order by sales desc;"
result = pd.read_sql_query(query, conn)

# Display the results
result

Month_name  year    sales
0     October  2023  21734.4
1      August  2022  16975.0
2       April  2023  10976.0
3    November  2022   9975.0
4    February  2023   9395.5
5     January  2023   8827.0
6        June  2023   8148.0
7    December  2023   6159.5
8        July  2022   4811.8
9   September  2023   4753.0
10        May  2022   4410.0
11      March  2022   4300.8

# **I) year-over-year sales to identify growth**

In [34]:
query = f"select Distinct year FROM cleanedorders1;"
result = pd.read_sql_query(query, conn)

# Display the results
result

year
0  2023
1  2022

**highest sales price**  - 2023 have highest sales price

In [35]:
query = f"SELECT year,sum(Selling_Price) as sales_price FROM cleanedorders1 group by year order by sales_price desc;"

result = pd.read_sql_query(query, conn)

# Display the results
result

year  sales_price
0  2023    1120268.5
1  2022    1095590.2

***highest sale month of 2023*** - based on selling price

In [36]:
query = f"SELECT year,COUNT(order_id) as count_of_orders,month_name,sum(Selling_Price) FROM cleanedorders1 where year='2023' and month='2';"
result = pd.read_sql_query(query, conn)

# Display the results
result

year  count_of_orders Month_name  sum(Selling_Price)
0  2023              401   February            128124.2

***highest orders placed on month of 2023*** - based on orders

In [37]:
query = f"SELECT year,COUNT(order_id) as count_of_orders,month_name,sum(Selling_Price) FROM cleanedorders1 where year='2023' and month='7';"
result = pd.read_sql_query(query, conn)

# Display the results
result

year  count_of_orders Month_name  sum(Selling_Price)
0  2023              477       July             90563.8

***highest profit on month of 2023*** - based on Profit

In [38]:
query = f"SELECT COUNT(order_id),month_name,sum(profit),sum(selling_price) FROM cleanedorders1 where year='2023' and month='10';"
result = pd.read_sql_query(query, conn)

# Display the results
result

COUNT(order_id) Month_name  sum(profit)  sum(selling_price)
0              418    October      12621.5            121061.5

**hint:-**

**highest selling price: - 2023:**

**highest profit : - 2022:**

**higest orders: - 2022:**

# **2023:**

***sum(Selling_Price) - 1120268.5 | sum(profit) - 103548.499 | count(order_id) - 4957***

# **2022:**

***sum(Selling_Price) - 1095590.2 |  sum(profit) - 101620.199 | count(order_id) - 5037***

# **II) year-over-year sales - decline in certain months**

In [39]:
query = f"SELECT Selling_Price,year,month_name,COUNT(order_id) as total_declin_orders FROM cleanedorders1 WHERE selling_price<=0 group by month_name order by total_declin_orders desc;"
result = pd.read_sql_query(query, conn)

# Display the results
result

Selling_Price  year Month_name  total_declin_orders
0             0.0  2022   November                   56
1             0.0  2023      April                   53
2             0.0  2023     August                   47
3             0.0  2022   December                   45
4             0.0  2022    January                   42
5             0.0  2022   February                   42
6             0.0  2022       June                   41
7             0.0  2022      March                   40
8             0.0  2023    October                   39
9             0.0  2022       July                   38
10            0.0  2022  September                   34
11            0.0  2022        May                   30

**total order decline both year of 2022 and 2023**

In [40]:

query = f"SELECT selling_Price,count(Order_Id) as total_order_Decline_both_years FROM cleanedorders1 WHERE selling_price<=0;"
result = pd.read_sql_query(query, conn)

# Display the results
result

Selling_Price  total_order_Decline_both_years
0            0.0                             507

# **Orders Decline of 2022**

In [41]:

query = f"SELECT Selling_Price,year,count(order_id) FROM cleanedorders1 WHERE selling_price<=0 and year='2022';"
result = pd.read_sql_query(query, conn)

# Display the results
result

Selling_Price  year  count(order_id)
0            0.0  2022              263

# **which month face highest decline of 2022**

In [42]:

query = f"SELECT Selling_Price,year,month_name, count(order_id) as counts FROM cleanedorders1 WHERE selling_price<=0 and year='2022' and month='11';"
result = pd.read_sql_query(query, conn)

# Display the results
result

Selling_Price  year Month_name  counts
0            0.0  2022   November      33

# **Orders decline of 2023**

In [43]:

query = f"SELECT Selling_Price,year, count(order_id) as counts FROM cleanedorders1 WHERE selling_price<=0 and year='2023';"
result = pd.read_sql_query(query, conn)

# Display the results
result

Selling_Price  year  counts
0            0.0  2023     244

# **which month face highest decline of 2023**

In [44]:

query = f"SELECT Selling_Price,year, count(order_id) as counts FROM cleanedorders1 WHERE selling_price<=0 and year='2023' and month='4';"
result = pd.read_sql_query(query, conn)

# Display the results
result

Selling_Price  year  counts
0            0.0  2023      31

# **III) Product Performance: **

In [45]:

query = f"SELECT row_number() over (order by product_id) as row, dense_rank() over (order by selling_price) as rank, product_id,quantity,max(profit),max(selling_price),min(discount) from cleanedorders1 group by product_id order by rank asc;"
result = pd.read_sql_query(query, conn)

# Display the results
result

row  rank       Product_Id  Quantity  max(profit)  max(selling_price)  \
0      137     1  FUR-FU-10000010         1          9.2                39.2   
1      138     1  FUR-FU-10000023         2          8.0                67.9   
2      144     1  FUR-FU-10000206         1          0.0                19.4   
3      158     1  FUR-FU-10000550         1          0.0                19.6   
4      193     1  FUR-FU-10001706         4          0.0                39.2   
...    ...   ...              ...       ...          ...                 ...   
1857  1421   241  OFF-SU-10000151         3        491.8              7780.5   
1858  1631   242  TEC-MA-10001047         4       1187.0              8827.0   
1859  1670   243  TEC-MA-10004125         4        560.0              7840.0   
1860  1615   244  TEC-CO-10004722         4       2975.0             16975.0   
1861  1643   245  TEC-MA-10002412         6       3624.4             21734.4   

      min(discount)  
0               0.0  
1               0.0  
2               0.0  
3               0.0  
4               0.0  
...             ...  
1857           98.2  
1858          260.0  
1859          160.0  
1860          224.0  
1861          905.6  

[1862 rows x 7 columns]

# **IV)Having**

In [46]:

query = f"SELECT row_number() over (order by product_id) as row, dense_rank() over (order by selling_price) as rank, product_id,quantity,max(profit),max(selling_price) as sale_price,min(discount) from cleanedorders1 group by product_id having max(selling_price)>=5000 order by sale_price asc;"
result = pd.read_sql_query(query, conn)

# Display the results
result

row  rank       Product_Id  Quantity  max(profit)  sale_price  \
0     6     2  TEC-CO-10001449         2        864.0      5184.0   
1     4     3  OFF-BI-10004995         4        652.4      5222.4   
2     3     1  OFF-BI-10003527         3        949.5      6159.5   
3     5     8  OFF-SU-10000151         3        491.8      7780.5   
4     8     6  TEC-MA-10000822         3        983.6      7833.6   
5    12    10  TEC-MA-10004125         4        560.0      7840.0   
6    10     5  TEC-MA-10001127         3        525.0      8575.0   
7     9     9  TEC-MA-10001047         4       1187.0      8827.0   
8     2     7  OFF-BI-10001120         3        851.0      9261.0   
9     1     4  OFF-BI-10000545         5        985.5      9395.5   
10    7    11  TEC-CO-10004722         4       2975.0     16975.0   
11   11    12  TEC-MA-10002412         6       3624.4     21734.4   

    min(discount)  
0            19.2  
1            32.7  
2            22.8  
3            98.2  
4            91.8  
5           160.0  
6            78.6  
7           260.0  
8            90.8  
9            22.8  
10          224.0  
11          905.6

# **V) Regional Sales Analysis:**

In [47]:

query = f"SELECT row_number() over (order by sum(profit)) as row, dense_rank() over (order by sum(selling_price)) as rank,region, count(quantity),sum(profit),sum(selling_price) as Total_Sales,sum(discount) from cleanedorders1 group by region order by Total_Sales desc;"
result = pd.read_sql_query(query, conn)

# Display the results
result

row  rank   Region  count(quantity)  sum(profit)  Total_Sales  \
0    4     4     West             3203      62558.6     699858.6   
1    3     3     East             2848      58220.0     654320.0   
2    2     2  Central             2323      46378.6     483668.6   
3    1     1    South             1620      38011.5     378011.5   

   sum(discount)  
0        25361.4  
1        24060.0  
2        17401.4  
3        13508.5

# **VI) Discount Analysis:**

# ** calculate the impact of discounts on sales**

**quantity_sold * (original_price - sales_price)**

orginal price is cost price

In [48]:

query = f"SELECT order_date,product_id, quantity, cost_price,list_price,discount,selling_price,profit, (quantity * (cost_price - selling_price)) AS impact FROM  cleanedorders1 order by impact desc;"
result = pd.read_sql_query(query, conn)

# Display the results
result

Order_Date       Product_Id  Quantity  cost_price  List_Price  Discount  \
0     2023-10-31  TEC-AC-10004227        10         100         100       4.0   
1     2023-01-21  FUR-TA-10003954         9        2540        2670     133.5   
2     2022-04-09  OFF-AP-10000358         7          90          90       4.5   
3     2023-07-06  OFF-PA-10002709        11          70          70       2.8   
4     2022-01-10  FUR-TA-10004767         6         290         300      15.0   
...          ...              ...       ...         ...         ...       ...   
9989  2022-10-16  TEC-MA-10000822         8        6850        8160     326.4   
9990  2023-01-03  TEC-MA-10001047         7        7640        9100     273.0   
9991  2023-02-25  OFF-BI-10000545        13        8410        9890     494.5   
9992  2022-08-14  TEC-CO-10004722         5       14000       17500     525.0   
9993  2023-10-13  TEC-MA-10002412         6       18110       22640     905.6   

      Selling_Price  Profit   impact  
0              96.0    -4.0     40.0  
1            2536.5    -3.5     31.5  
2              85.5    -4.5     31.5  
3              67.2    -2.8     30.8  
4             285.0    -5.0     30.0  
...             ...     ...      ...  
9989         7833.6   983.6  -7868.8  
9990         8827.0  1187.0  -8309.0  
9991         9395.5   985.5 -12811.5  
9992        16975.0  2975.0 -14875.0  
9993        21734.4  3624.4 -21746.4  

[9994 rows x 9 columns]

# **Identify products with discounts greater than 20%**

dicount_percent = (cost_price - selling_price) / cost_price


In [49]:

query = f"SELECT product_id,selling_price,discount,discount_percent FROM cleanedorders1 WHERE discount_percent > 0.20 order by discount_percent asc;"
result = pd.read_sql_query(query, conn)

# Display the results
result

Product_Id  Selling_Price  Discount  Discount_Percent
0     FUR-BO-10001798          254.8       5.2                 2
1     FUR-TA-10000577          940.8      19.2                 2
2     OFF-BI-10003910           19.6       0.4                 2
3     OFF-BI-10003656          401.8       8.2                 2
4     OFF-BI-10002215           19.6       0.4                 2
...               ...            ...       ...               ...
9989  OFF-AP-10003622           19.0       1.0                 5
9990  TEC-PH-10002496          237.5      12.5                 5
9991  OFF-LA-10004484           19.0       1.0                 5
9992  TEC-AC-10003399           38.0       2.0                 5
9993  TEC-PH-10004006          199.5      10.5                 5

[9994 rows x 4 columns]

# **1.# Find top 10 highest revenue generating products**

In [50]:

query = f"SELECT product_id,selling_price from cleanedorders1 order by selling_price desc LIMIT 10;"
result = pd.read_sql_query(query, conn)

# Display the results
result

Product_Id  Selling_Price
0  TEC-MA-10002412        21734.4
1  TEC-CO-10004722        16975.0
2  TEC-CO-10004722        13440.0
3  TEC-CO-10004722        10976.0
4  TEC-CO-10004722         9975.0
5  OFF-BI-10000545         9395.5
6  OFF-BI-10001120         9261.0
7  TEC-MA-10001047         8827.0
8  TEC-MA-10001127         8575.0
9  TEC-CO-10004722         8148.0


# **2. # Find the top 5 cities with the highest profit margins**

In [51]:

query = f"SELECT city,sum(profit) as total_profit from cleanedorders1 group by city order by total_profit  desc limit 5;"
result = pd.read_sql_query(query, conn)

# Display the results
result

City  total_profit
0  New York City       20925.7
1    Los Angeles       15088.4
2  San Francisco       10140.1
3        Seattle        9828.7
4   Philadelphia        9788.3


# **3.# Calculate the total discount given for each category**

In [52]:

query = f"SELECT category,sum(discount) as discount_price from cleanedorders1 group by category;"
result = pd.read_sql_query(query, conn)

# Display the results
result

Category  discount_price
0        Furniture         26049.3
1  Office Supplies         24935.5
2       Technology         29346.5

# **4.# Find the average sale price per product category**

In [53]:

query = f"SELECT category,avg(selling_price) from cleanedorders1 group by category;"
result = pd.read_sql_query(query, conn)

# Display the results
result

Category  avg(selling_price)
0        Furniture          337.265771
1  Office Supplies          115.108613
2       Technology          436.856253

# **5.# Find the region with the highest average sale price**


In [54]:
query = f"SELECT region,avg(selling_price) as avgsale from cleanedorders1 group by region order by avgsale desc;"
result = pd.read_sql_query(query, conn)

# Display the results
result

Region     avgsale
0    South  233.340432
1     East  229.747191
2     West  218.500968
3  Central  208.208610


# **6.# Find the total profit per category**

In [55]:
query = f"SELECT category,sum(profit) from cleanedorders1 group by category;"
result = pd.read_sql_query(query, conn)

# Display the results
result

Category  sum(profit)
0        Furniture      66480.7
1  Office Supplies      62254.5
2       Technology      76433.5

# **7.# Identify the top 3 segments with the highest quantity of orders**


In [56]:
query = f"SELECT segment,count(order_id) as orders from cleanedorders1 group by segment order by orders desc LIMIT 3;"
result = pd.read_sql_query(query, conn)

# Display the results
result

Segment  orders
0     Consumer    5191
1    Corporate    3020
2  Home Office    1783

# **8.# Determine the average discount percentage given per region**

In [57]:
query = f"SELECT region,avg(discount_percent) as disc from cleanedorders1 group by region order by disc desc;"
result = pd.read_sql_query(query, conn)

# Display the results
result

Region      disc
0     East  3.493680
1     West  3.484858
2    South  3.482716
3  Central  3.472234

# **9.# Find the product category with the highest total profit**


In [58]:
query = f"SELECT category,sum(profit) as total_profit from cleanedorders1 group by category ORDER by total_profit desc;"
result = pd.read_sql_query(query, conn)

# Display the results
result

Category  total_profit
0       Technology       76433.5
1        Furniture       66480.7
2  Office Supplies       62254.5

# **10.# Calculate the total revenue generated per year**


In [59]:
query = f"SELECT year,sum(selling_price) as Revenue from cleanedorders1 group by year ORDER by Revenue desc;"
result = pd.read_sql_query(query, conn)

# Display the results
result

year    Revenue
0  2023  1120268.5
1  2022  1095590.2

# **Second Half Questions**

In [60]:
# second half questions

In [61]:
df.head(5)

Unnamed: 0  Order_Id  Order_Date       Ship_Mode    Segment        Country  \
0           0         1  2023-03-01    Second Class   Consumer  United States   
1           1         2  2023-08-15    Second Class   Consumer  United States   
2           2         3  2023-01-10    Second Class  Corporate  United States   
3           3         4  2022-06-18  Standard Class   Consumer  United States   
4           4         5  2022-07-13  Standard Class   Consumer  United States   

              City       State  Postal_Code Region  ... List_Price Quantity  \
0        Henderson    Kentucky        42420  South  ...        260        2   
1        Henderson    Kentucky        42420  South  ...        730        3   
2      Los Angeles  California        90036   West  ...         10        2   
3  Fort Lauderdale     Florida        33311  South  ...        960        5   
4  Fort Lauderdale     Florida        33311  South  ...         20        2   

  Discount_Percent  Selling_Price  Discount  Profit  year  month  day  \
0                2          254.8       5.2    14.8  2023      3    1   
1                3          708.1      21.9   108.1  2023      8   15   
2                5            9.5       0.5    -0.5  2023      1   10   
3                2          940.8      19.2   160.8  2022      6   18   
4                5           19.0       1.0    -1.0  2022      7   13   

   Month_name  
0       March  
1      August  
2     January  
3        June  
4        July  

[5 rows x 24 columns]

In [62]:
df1=df[['Order_Id','Order_Date','Ship_Mode','Segment','Country','City','State','Postal_Code','Region','Category','Sub_Category']]
df1.head(3)

Order_Id  Order_Date     Ship_Mode    Segment        Country         City  \
0         1  2023-03-01  Second Class   Consumer  United States    Henderson   
1         2  2023-08-15  Second Class   Consumer  United States    Henderson   
2         3  2023-01-10  Second Class  Corporate  United States  Los Angeles   

        State  Postal_Code Region         Category Sub_Category  
0    Kentucky        42420  South        Furniture    Bookcases  
1    Kentucky        42420  South        Furniture       Chairs  
2  California        90036   West  Office Supplies       Labels

In [63]:
df2=df[['Product_Id','cost_price','List_Price','Quantity','Discount_Percent','Selling_Price','Discount','Profit','year','month','day','Month_name']]
df2.head(3)

Product_Id  cost_price  List_Price  Quantity  Discount_Percent  \
0  FUR-BO-10001798         240         260         2                 2   
1  FUR-CH-10000454         600         730         3                 3   
2  OFF-LA-10000240          10          10         2                 5   

   Selling_Price  Discount  Profit  year  month  day Month_name  
0          254.8       5.2    14.8  2023      3    1      March  
1          708.1      21.9   108.1  2023      8   15     August  
2            9.5       0.5    -0.5  2023      1   10    January

# **next 11 to 20 questions**

# **11) whats is primary key & Foreign Key**

# ***Primary Key***

primary key is major column of table its not create dublicate value mostly primary key used as a ID with not null option and somtimes add autoincrement options on sql table

*if you want to drop only primary key option not column

before create a primary key give constraint

eg:- alter table df1 CONSTRAINT PK_orders primary key(order_Id)

else

its enough ->  alter table df1 primary key(order_Id)

# ***Foreign key***

foreign key is child key its depended on primary key based

eg: star schmea
primary key in 1st table and foreign key in 2nd table,
its act like on connecting a bridge of both tables

when create a new row on 2nd table but value of foreign key not in primary key unable to create new row on 2nd table

eg: department in college they give primary value column of department as department id. students in college they also create primary key as student id and name, details, and connect to department to avoid duplicate and fast loading of date or easily update of database, create as foreign key connect to the primary key of department

In [64]:


# 1) ALTER TABLE df1 ADD PRIMARY KEY (order_id);

#  ALTER TABLE df2 ADD COLUMN order_id INT;

# ALTER TABLE df2 ADD FOREIGN KEY(order_id) REFERENCES df1(order_id);

# **12) which region city have poor performance**

**a) which region, city poor performence of sales**


In [65]:
query = f"SELECT region,city,min(selling_price) as sales from cleanedorders1 GROUP by city order by sales asc;"
result = pd.read_sql_query(query, conn)

# Display the results
result

Region          City   sales
0    Central       Abilene     0.0
1       East         Akron     0.0
2       West   Albuquerque     0.0
3       East     Allentown     0.0
4       East       Altoona     0.0
..       ...           ...     ...
526     West      Missoula   480.2
527  Central  Saint Peters   679.0
528    South     Kissimmee   712.5
529  Central  Independence   823.2
530     West      Cheyenne  1536.0

[531 rows x 3 columns]

**b) which region, city poor performence of profit**

In [66]:
query = f"SELECT region,city,min(profit) as prof from cleanedorders1 GROUP by city order by prof asc;"
result = pd.read_sql_query(query, conn)

# Display the results
result

Region          City   prof
0       West     Coachella   -5.0
1    Central        Dallas   -5.0
2    Central    Fort Worth   -5.0
3    Central       Houston   -5.0
4      South    Louisville   -5.0
..       ...           ...    ...
526  Central        Normal   51.5
527     West      Whittier   56.8
528  Central  Saint Peters   59.0
529     West      Missoula   70.2
530     West      Cheyenne  156.0

[531 rows x 3 columns]

**c) which region,city poor performence of orders**


In [67]:
query = f"SELECT region,city,min(quantity) as qty from cleanedorders1 GROUP by city order by qty asc;"
result = pd.read_sql_query(query, conn)

# Display the results
result

Region         City  qty
0       West  Albuquerque    1
1      South   Alexandria    1
2    Central     Amarillo    1
3       West      Anaheim    1
4       East      Andover    1
..       ...          ...  ...
526  Central  Tinley Park    7
527    South     Bartlett    8
528  Central       Conroe    9
529    South       Rogers    9
530     West    Vacaville   11

[531 rows x 3 columns]

# **13) split two column as df1 and df2 and upload as sql database | create new colum for df2 2nd table as foreigh key**



In [68]:
df.head(1)

Unnamed: 0  Order_Id  Order_Date     Ship_Mode   Segment        Country  \
0           0         1  2023-03-01  Second Class  Consumer  United States   

        City     State  Postal_Code Region  ... List_Price Quantity  \
0  Henderson  Kentucky        42420  South  ...        260        2   

  Discount_Percent  Selling_Price  Discount  Profit  year  month  day  \
0                2          254.8       5.2    14.8  2023      3    1   

   Month_name  
0       March  

[1 rows x 24 columns]

In [69]:
df2['Order_Id'] = df1.assign(Order_Id=df1['Order_Id'].values)['Order_Id']

<ipython-input-69-93f811f8e4b2>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df2['Order_Id'] = df1.assign(Order_Id=df1['Order_Id'].values)['Order_Id']


In [70]:
df2.head(3)

Product_Id  cost_price  List_Price  Quantity  Discount_Percent  \
0  FUR-BO-10001798         240         260         2                 2   
1  FUR-CH-10000454         600         730         3                 3   
2  OFF-LA-10000240          10          10         2                 5   

   Selling_Price  Discount  Profit  year  month  day Month_name  Order_Id  
0          254.8       5.2    14.8  2023      3    1      March         1  
1          708.1      21.9   108.1  2023      8   15     August         2  
2            9.5       0.5    -0.5  2023      1   10    January         3

In [71]:
df1.head(3)

Order_Id  Order_Date     Ship_Mode    Segment        Country         City  \
0         1  2023-03-01  Second Class   Consumer  United States    Henderson   
1         2  2023-08-15  Second Class   Consumer  United States    Henderson   
2         3  2023-01-10  Second Class  Corporate  United States  Los Angeles   

        State  Postal_Code Region         Category Sub_Category  
0    Kentucky        42420  South        Furniture    Bookcases  
1    Kentucky        42420  South        Furniture       Chairs  
2  California        90036   West  Office Supplies       Labels

In [72]:
df1.to_csv('df1.csv')

In [73]:
df2.to_csv('df2.csv')

In [74]:
import sqlite3
import pandas as pd

# File name of the uploaded CSV
csv_filename1 = "/content/df1.csv"  # Replace this with the uploaded file name
csv_filename2 = "/content/df2.csv"  # This should point to the correct CSV file for df2

# Load CSV into a pandas DataFrame
df1 = pd.read_csv(csv_filename1)
df2 = pd.read_csv(csv_filename2) # Load data from the correct file for df2

# Connect to SQLite database (or create a new one)
conn = sqlite3.connect("example.db")
cursor = conn.cursor()

# Write DataFrame to SQLite table
table_name1 = "df1"  # Specify your table name
df1.to_sql(table_name1, conn, if_exists="replace", index=False)

table_name2 = "df2"  # Specify your table name
df2.to_sql(table_name2, conn, if_exists="replace", index=False)

print(f"Table '{table_name1}' created in SQLite database.")
print(f"Table '{table_name2}' created in SQLite database.")

Table 'df1' created in SQLite database.
Table 'df2' created in SQLite database.


In [75]:
# df2['Order_Id'] = df1.assign(Order_Id=df1['Order_Id'].values)['Order_Id']

# 1) ALTER TABLE df1 ADD PRIMARY KEY (order_id);
# 1) ALTER TABLE df1 constraints PK_order ADD PRIMARY KEY (order_id);

#  ALTER TABLE df2 ADD COLUMN order_id INT;

# ALTER TABLE df2 ADD FOREIGN KEY(order_id) REFERENCES df1(order_id);
# or
# ALTER TABLE df2 constraints FK_order ADD FOREIGN KEY(order_id) REFERENCES df1(order_id);

constraints FK_order - > its usefull when drop status of foreign key of that column otherwise can't delete that column because that primary key column wont allowed

# **14. join df1 and df2 two table using order id find as a discount from table 1 and city from table 2**



In [76]:
df1=pd.read_csv('/content/df1.csv')
df2=pd.read_csv('/content/df2.csv')

In [77]:
df1.head(2)

Unnamed: 0  Order_Id  Order_Date     Ship_Mode   Segment        Country  \
0           0         1  2023-03-01  Second Class  Consumer  United States   
1           1         2  2023-08-15  Second Class  Consumer  United States   

        City     State  Postal_Code Region   Category Sub_Category  
0  Henderson  Kentucky        42420  South  Furniture    Bookcases  
1  Henderson  Kentucky        42420  South  Furniture       Chairs

In [78]:
df2.head(2)

Unnamed: 0       Product_Id  cost_price  List_Price  Quantity  \
0           0  FUR-BO-10001798         240         260         2   
1           1  FUR-CH-10000454         600         730         3   

   Discount_Percent  Selling_Price  Discount  Profit  year  month  day  \
0                 2          254.8       5.2    14.8  2023      3    1   
1                 3          708.1      21.9   108.1  2023      8   15   

  Month_name  Order_Id  
0      March         1  
1     August         2

select city from 1st table then take discount percent from 2nd table then from 1st table join 2nd table based on 1st primary key equal of 2nd table of foreign key thats all

In [79]:
query = f"SELECT df1.City,df2.Discount_Percent FROM df1 JOIN df2 ON df1.Order_Id = df2.Order_Id;"
result = pd.read_sql_query(query, conn)

# Display the results
result

City  Discount_Percent
0           Henderson                 2
1           Henderson                 3
2         Los Angeles                 5
3     Fort Lauderdale                 2
4     Fort Lauderdale                 5
...               ...               ...
9989            Miami                 4
9990       Costa Mesa                 4
9991       Costa Mesa                 2
9992       Costa Mesa                 3
9993      Westminster                 3

[9994 rows x 2 columns]

SELECT df2.order_id,df2.Product_Id FROM df2 join df1 on df2.Order_Id = df1.Order_Id;


SELECT df1.city, df2.Discount_Percent
FROM df1
JOIN df2
ON df1.order_id = df2.order_id;

select city from 1st table then take discount percent from 2nd table then from 1st table join 2nd table based on 1st primary key equal of 2nd table of foreign key thats all

# **15. nested querry or sub querry**

select product_id from df2 where order_id in ( select order_id from df1 where order_id = '5')


In [80]:
query = f"select product_id,Selling_Price from df2 where order_id in (select order_id from df1 where order_id = '5')"
result = pd.read_sql_query(query, conn)

# Display the results
result

Product_Id  Selling_Price
0  OFF-ST-10000760           19.0

select selling_price,discount_percent,cost_price from df2 where order_id in ( select order_id from df1 where order_id = '20')


# **16.left join and right join**



eg: left join focus and marked on right table missing or null values

right join focus and marked on left table missing or null values

# **17.Seasonality:**

****Are there any seasonal trends in the data, such as higher sales during holidays or specific months?*****

SELECT month_name,year,day,sum(quantity) from cleanedorders1 GROUP by month_name order by quantity desc;



In [81]:
query = f"SELECT month_name,year,day,sum(quantity) as qty from cleanedorders1 GROUP by month_name order by qty desc;"
result = pd.read_sql_query(query, conn)

# Display the results
result

Month_name  year  day   qty
0        July  2022   13  3423
1     October  2022    5  3371
2      August  2023   15  3348
3    December  2022   28  3249
4         May  2023   16  3200
5     January  2023   10  3193
6       April  2022    5  3149
7       March  2023    1  3098
8    February  2022   11  3085
9    November  2023    9  3055
10       June  2022   18  2950
11  September  2022   28  2752

# **seasonality based profit**

In [82]:
query = f"SELECT month_name,year,day,sum(profit) as profit from cleanedorders1 GROUP by month_name order by profit desc;"
result = pd.read_sql_query(query, conn)

# Display the results
result

Month_name  year  day   profit
0     October  2022    5  22254.2
1    February  2022   11  20655.2
2      August  2023   15  20121.6
3       April  2022    5  18830.2
4    December  2022   28  18016.0
5     January  2023   10  17365.1
6        July  2022   13  15586.0
7         May  2023   16  15276.2
8       March  2023    1  14718.3
9        June  2022   18  14487.0
10   November  2023    9  14288.1
11  September  2022   28  13570.8

# **18. Segment:**

# **a) maximum sales of each segment**

In [83]:
query = f"SELECT segment,max(selling_price) as sales from cleanedorders1 GROUP by segment order by sales desc;"
result = pd.read_sql_query(query, conn)

# Display the results
result

Segment    sales
0  Home Office  21734.4
1    Corporate  16975.0
2     Consumer  13440.0

# **b) total sales of each segment**

In [84]:
query = f"SELECT segment,sum(selling_price) as sales from cleanedorders1 GROUP by segment order by sales desc;"
result = pd.read_sql_query(query, conn)

# Display the results
result

Segment      sales
0     Consumer  1119696.0
1    Corporate   681657.7
2  Home Office   414505.0

# **c)total profit of each segment**

In [85]:
query = f"SELECT segment,sum(Profit) as profit from cleanedorders1 GROUP by segment order by profit desc;"
result = pd.read_sql_query(query, conn)

# Display the results
result

Segment    profit
0     Consumer  101586.0
1    Corporate   65637.7
2  Home Office   37945.0

# **d) total quantaty of each segment**

In [86]:
query = f"SELECT segment,sum(quantity) as qty from cleanedorders1 GROUP by segment order by qty desc;"
result = pd.read_sql_query(query, conn)

# Display the results
result

Segment    qty
0     Consumer  19521
1    Corporate  11608
2  Home Office   6744

# **19. Discount Effectiveness:**

**discounts or promotions impact sales volume and profitability**

# **min profit for discount**

In [87]:
query = f"SELECT product_id,order_date,city,category,sub_category,quantity,selling_price,min(profit) as profit, (list_price - selling_price ) / selling_price * 100 AS discount FROM cleanedorders1 WHERE selling_price IS NOT NULL GROUP by quantity ORDER BY profit<0 asc"
result = pd.read_sql_query(query, conn)

# Display the results
result

Product_Id  Order_Date            City         Category Sub_Category  \
0   TEC-PH-10002597  2023-06-30      Fort Worth       Technology       Phones   
1   OFF-ST-10004804  2023-02-17    Philadelphia  Office Supplies      Storage   
2   TEC-PH-10004165  2022-02-01    Philadelphia       Technology       Phones   
3   TEC-PH-10003215  2022-01-22   San Francisco       Technology       Phones   
4   FUR-CH-10001146  2022-09-11      Louisville        Furniture       Chairs   
5   FUR-TA-10004767  2022-01-10          Dallas        Furniture       Tables   
6   OFF-AP-10000358  2022-04-09     Los Angeles  Office Supplies   Appliances   
7   OFF-PA-10001509  2023-06-11  Salt Lake City  Office Supplies        Paper   
8   FUR-TA-10003954  2023-01-21      Bellingham        Furniture       Tables   
9   TEC-AC-10004227  2023-10-31           Salem       Technology  Accessories   
10  OFF-PA-10002709  2023-07-06   New York City  Office Supplies        Paper   
11  OFF-FA-10003495  2023-10-19         Houston  Office Supplies    Fasteners   
12  OFF-BI-10003984  2023-09-16   New York City  Office Supplies      Binders   
13  OFF-AR-10003631  2022-11-01         Chicago  Office Supplies          Art   

    Quantity  Selling_Price  profit  discount  
0          1           95.0    -5.0  5.263158  
1          2           76.0    -4.0  5.263158  
2          3          465.5    -4.5  5.263158  
3          4           95.0    -5.0  5.263158  
4          5          285.0    -5.0  5.263158  
5          6          285.0    -5.0  5.263158  
6          7           85.5    -4.5  5.263158  
7          8           66.5    -3.5  5.263158  
8          9         2536.5    -3.5  5.263158  
9         10           96.0    -4.0  4.166667  
10        11           67.2    -2.8  4.166667  
11        12           57.6    -2.4  4.166667  
12        13          288.0    -2.0  4.166667  
13        14           48.0    -2.0  4.166667

# **max profit for discount**

In [88]:
query = f"SELECT product_id,order_date,city,category,sub_category,quantity,selling_price,max(profit) as profit, (list_price - selling_price ) / selling_price * 100 AS discount FROM cleanedorders1 WHERE selling_price GROUP by quantity ORDER BY profit>0 DESC"
result = pd.read_sql_query(query, conn)

# Display the results
result

Product_Id  Order_Date           City         Category Sub_Category  \
0   OFF-AP-10002684  2022-09-06  San Francisco  Office Supplies   Appliances   
1   OFF-BI-10000545  2023-02-25        Detroit  Office Supplies      Binders   
2   FUR-TA-10001095  2022-08-29      San Diego        Furniture       Tables   
3   FUR-CH-10002073  2022-02-07  New York City        Furniture       Chairs   
4   TEC-MA-10000112  2023-05-03         Sparks       Technology     Machines   
5   TEC-CO-10001449  2022-08-05     Providence       Technology      Copiers   
6   TEC-MA-10000822  2022-10-16    San Antonio       Technology     Machines   
7   TEC-MA-10001047  2023-01-03       Lakewood       Technology     Machines   
8   TEC-MA-10002412  2023-10-13   Jacksonville       Technology     Machines   
9   TEC-CO-10004722  2022-08-14      Lafayette       Technology      Copiers   
10  TEC-CO-10004722  2023-06-07   Philadelphia       Technology      Copiers   
11  TEC-CO-10004722  2022-11-29         Newark       Technology      Copiers   
12  TEC-MA-10000418  2023-07-30  San Francisco       Technology     Machines   
13  OFF-BI-10003527  2022-12-22    Los Angeles  Office Supplies      Binders   

    Quantity  Selling_Price  profit  discount  
0         14         1666.0   256.0  2.040816  
1         13         9395.5   985.5  5.263158  
2         12         1586.5   236.5  5.263158  
3         11         2567.6   467.6  2.040816  
4         10         2304.0   334.0  4.166667  
5          9         5184.0   864.0  4.166667  
6          8         7833.6   983.6  4.166667  
7          7         8827.0  1187.0  3.092784  
8          6        21734.4  3624.4  4.166667  
9          5        16975.0  2975.0  3.092784  
10         4         8148.0   838.0  3.092784  
11         3         9975.0   735.0  5.263158  
12         2         4704.0   674.0  2.040816  
13         1          989.4   169.4  3.092784

# **quantaty increase for discount**


In [89]:
query = f"SELECT product_id,order_date,city,category,sub_category,quantity,selling_price,profit, (list_price - selling_price ) / selling_price * 100 AS discount FROM cleanedorders1 WHERE selling_price IS NOT NULL GROUP by quantity ORDER BY quantity>0 DESC"
result = pd.read_sql_query(query, conn)

# Display the results
result

Product_Id  Order_Date             City         Category  \
0   OFF-FA-10000621  2023-06-06         Columbus  Office Supplies   
1   TEC-AC-10003832  2022-05-28        San Diego       Technology   
2   TEC-PH-10000702  2023-07-13     Philadelphia       Technology   
3   TEC-PH-10000215  2022-06-12         Franklin       Technology   
4   OFF-LA-10001074  2023-10-03     Philadelphia  Office Supplies   
5   FUR-TA-10001539  2023-03-31      Los Angeles        Furniture   
6   OFF-AR-10002671  2022-04-10          Gilbert  Office Supplies   
7   FUR-FU-10001487  2022-03-13      Los Angeles        Furniture   
8   TEC-PH-10002275  2022-01-25      Los Angeles       Technology   
9   FUR-TA-10000577  2022-06-18  Fort Lauderdale        Furniture   
10  OFF-AR-10002833  2022-12-28      Los Angeles  Office Supplies   
11  FUR-CH-10000454  2023-08-15        Henderson        Furniture   
12  FUR-BO-10001798  2023-03-01        Henderson        Furniture   
13  FUR-CH-10003061  2022-01-31       New Albany        Furniture   

   Sub_Category  Quantity  Selling_Price  Profit  discount  
0     Fasteners        14           39.2    -0.8  2.040816  
1   Accessories        13         3216.0   166.0  4.166667  
2        Phones        12           67.2     7.2  4.166667  
3        Phones        11          364.8    14.8  4.166667  
4        Labels        10           98.0     8.0  2.040816  
5        Tables         9         1658.7   188.7  3.092784  
6           Art         8         1076.7   136.7  3.092784  
7   Furnishings         7           48.5    -1.5  3.092784  
8        Phones         6          864.5     4.5  5.263158  
9        Tables         5          940.8   160.8  2.040816  
10          Art         4            9.7    -0.3  3.092784  
11       Chairs         3          708.1   108.1  3.092784  
12    Bookcases         2          254.8    14.8  2.040816  
13       Chairs         1           85.5    15.5  5.263158

# **quantaty increase for discount using having method**


In [90]:
query = f"SELECT product_id,order_date,city,category,sub_category,quantity,selling_price,profit, (list_price - selling_price ) / selling_price * 100 AS discount FROM cleanedorders1 WHERE selling_price IS NOT NULL GROUP by quantity having quantity>0 order by quantity desc"
result = pd.read_sql_query(query, conn)

# Display the results
result

Product_Id  Order_Date             City         Category  \
0   OFF-FA-10000621  2023-06-06         Columbus  Office Supplies   
1   TEC-AC-10003832  2022-05-28        San Diego       Technology   
2   TEC-PH-10000702  2023-07-13     Philadelphia       Technology   
3   TEC-PH-10000215  2022-06-12         Franklin       Technology   
4   OFF-LA-10001074  2023-10-03     Philadelphia  Office Supplies   
5   FUR-TA-10001539  2023-03-31      Los Angeles        Furniture   
6   OFF-AR-10002671  2022-04-10          Gilbert  Office Supplies   
7   FUR-FU-10001487  2022-03-13      Los Angeles        Furniture   
8   TEC-PH-10002275  2022-01-25      Los Angeles       Technology   
9   FUR-TA-10000577  2022-06-18  Fort Lauderdale        Furniture   
10  OFF-AR-10002833  2022-12-28      Los Angeles  Office Supplies   
11  FUR-CH-10000454  2023-08-15        Henderson        Furniture   
12  FUR-BO-10001798  2023-03-01        Henderson        Furniture   
13  FUR-CH-10003061  2022-01-31       New Albany        Furniture   

   Sub_Category  Quantity  Selling_Price  Profit  discount  
0     Fasteners        14           39.2    -0.8  2.040816  
1   Accessories        13         3216.0   166.0  4.166667  
2        Phones        12           67.2     7.2  4.166667  
3        Phones        11          364.8    14.8  4.166667  
4        Labels        10           98.0     8.0  2.040816  
5        Tables         9         1658.7   188.7  3.092784  
6           Art         8         1076.7   136.7  3.092784  
7   Furnishings         7           48.5    -1.5  3.092784  
8        Phones         6          864.5     4.5  5.263158  
9        Tables         5          940.8   160.8  2.040816  
10          Art         4            9.7    -0.3  3.092784  
11       Chairs         3          708.1   108.1  3.092784  
12    Bookcases         2          254.8    14.8  2.040816  
13       Chairs         1           85.5    15.5  5.263158

# **20) ship mode prefered most**


In [91]:
query = f"SELECT ship_mode,sum(quantity) as qty from cleanedorders1 GROUP by ship_mode order by qty desc;"
result = pd.read_sql_query(query, conn)

# Display the results
result

Ship_Mode    qty
0  Standard Class  22768
1    Second Class   7423
2     First Class   5693
3        Same Day   1960
4   Not Available     23
5         unknown      5
6               0      1

# **Streamlit**

In [92]:
# streamlit

In [93]:
!pip install -q streamlit

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 44.3/44.3 kB 1.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 9.1/9.1 MB 47.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.9/6.9 MB 58.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 79.1/79.1 kB 4.5 MB/s eta 0:00:00


In [94]:
ipv4 = !curl ipv4.icanhazip.com
ipv4

['34.139.73.38']

In [95]:
%%writefile streamlit5_app.py

import streamlit as st
import sqlite3
import pandas as pd

# Streamlit app title
st.title("Retail Order Data Analysis (1st-Project)")

# File uploader in Streamlit
# File name of the uploaded CSV (replace with the actual uploaded file)
csv_filename = "/content/cleaned-orders-1.csv"  # Replace with the uploaded file name
csv_filename1 = "/content/df1.csv"
csv_filename2 = "/content/df2.csv"

# Load CSV into a pandas DataFrame
df = pd.read_csv(csv_filename)
df1 = pd.read_csv(csv_filename1)
df2 = pd.read_csv(csv_filename2)

# Connect to SQLite database (or create a new one)
conn = sqlite3.connect("example.db")
cursor = conn.cursor()

# Write DataFrame to SQLite table (replace with your table name)
table_name = "cleanedorders1"
table_name1 = "df1"
table_name2 = "df2"

df.to_sql(table_name, conn, if_exists="replace", index=False)
df1.to_sql(table_name1, conn, if_exists="replace", index=False)
df2.to_sql(table_name2, conn, if_exists="replace", index=False)

st.success(f"Tables '{table_name}', '{table_name1}', and '{table_name2}' created in SQLite database.")


# Define two sets of queries as dictionaries with descriptive names
query_sets = {
    "Sales Analysis Queries": {
        "Top-Selling Products": "SELECT row_number() over(order by product_id) as row, product_id, selling_price FROM cleanedorders1 ORDER BY selling_price DESC;",
        "Monthly Sales Analysis": "SELECT month_name, year, max(Selling_Price) as sales FROM cleanedorders1 GROUP BY month ORDER BY sales DESC;",
        "Product Performance - using Group By Method": "SELECT row_number() over (order by product_id) as row, dense_rank() over (order by selling_price) as rank, product_id, quantity, max(profit), max(selling_price), min(discount) FROM cleanedorders1 GROUP BY product_id ORDER BY rank ASC;",
        "Product Performance - using Group By and Having Method": "SELECT row_number() over (order by product_id) as row, dense_rank() over (order by selling_price) as rank, product_id,quantity,max(profit),max(selling_price) as sale_price,min(discount) from cleanedorders1 group by product_id having max(selling_price)>=5000 order by sale_price asc;",
    },
    "Impact Analysis Queries": {
        "Regional Sales Analysis": "SELECT row_number() over (order by sum(profit)) as row, dense_rank() over (order by sum(selling_price)) as rank, region, count(quantity), sum(profit), sum(selling_price) as Total_Sales, sum(discount) FROM cleanedorders1 GROUP BY region ORDER BY Total_Sales DESC;",
        "Discount Analysis": "SELECT order_date, product_id, quantity, cost_price, list_price, discount, selling_price, profit, (quantity * (cost_price - selling_price)) AS impact FROM cleanedorders1 ORDER BY impact DESC;",
    },
    "year-over-year sales to identify growth": {
        "highest sales price of year": "SELECT year,sum(Selling_Price) as sales_price FROM cleanedorders1 group by year order by sales_price desc;",
        "highest sale month of 2023 ": "SELECT year,COUNT(order_id) as count_of_orders,month_name,sum(Selling_Price) FROM cleanedorders1 where year='2023' and month='2';",
        "highest orders placed on month of 2023 ": "SELECT year,COUNT(order_id) as count_of_orders,month_name,sum(Selling_Price) FROM cleanedorders1 where year='2023' and month='7';",
        "highest profit on month of 2023": "SELECT COUNT(order_id),month_name,sum(profit),sum(selling_price) FROM cleanedorders1 where year='2023' and month='10';",
    },
    "year-over-year sales - decline in certain months": {
        "decline sales month of year": "SELECT Selling_Price,year,month_name,COUNT(order_id) as total_declin_orders FROM cleanedorders1 WHERE selling_price<=0 group by month_name order by total_declin_orders desc;",
        "total order decline both year of 2022 and 2023": "SELECT selling_Price,count(Order_Id) as total_order_Decline_both_years FROM cleanedorders1 WHERE selling_price<=0;",
        "Orders Decline of 2022 ": "SELECT Selling_Price,year,count(order_id) FROM cleanedorders1 WHERE selling_price<=0 and year='2022';",
        "which month face highest decline of 2022": "SELECT Selling_Price,year,month_name, count(order_id) as counts FROM cleanedorders1 WHERE selling_price<=0 and year='2022' and month='11';",
        "Orders Decline of 2023 ": "SELECT Selling_Price,year,count(order_id) FROM cleanedorders1 WHERE selling_price<=0 and year='2023';",
        "which month face highest decline of 2023": "SELECT Selling_Price,year,month_name, count(order_id) as counts FROM cleanedorders1 WHERE selling_price<=0 and year='2023' and month='3';",

    },
     "First 10 Questions": {
        "1.# Find top 10 highest revenue generating products": "SELECT product_id,selling_price from cleanedorders1 order by selling_price desc LIMIT 10;",
        "2.# Find the top 5 cities with the highest profit margins": "SELECT city,sum(profit) as total_profit from cleanedorders1 group by city order by total_profit  desc limit 5;",
        "3.# Calculate the total discount given for each category": "SELECT category,sum(discount) as discount_price from cleanedorders1 group by category;",
        "4.# Find the average sale price per product category": "SELECT category,avg(selling_price) from cleanedorders1 group by category;",
        "5.# Find the region with the highest average sale price": "SELECT region,avg(selling_price) as avgsale from cleanedorders1 group by region order by avgsale desc;",
        "6.# Find the total profit per category": "SELECT category,sum(profit) from cleanedorders1 group by category;",
        "7.# Identify the top 3 segments with the highest quantity of orders": "SELECT segment,count(order_id) as orders from cleanedorders1 group by segment order by orders desc LIMIT 3;",
        "8.# Determine the average discount percentage given per region ": "SELECT region,avg(discount_percent) as disc from cleanedorders1 group by region order by disc desc;",
        "9.# Find the product category with the highest total profit ": "SELECT category,sum(profit) as total_profit from cleanedorders1 group by category ORDER by total_profit desc;",
        "10.# Calculate the total revenue generated per year": "SELECT year,sum(selling_price) as Revenue from cleanedorders1 group by year ORDER by Revenue desc;",
    },
    "11 to 20 Questions using primary and foreign key join table process": {
        "11.# which region city have poor performance": "SELECT region,city,min(selling_price) as sales from cleanedorders1 GROUP by city order by sales asc;",
        "12.# join df1 and df2 two table using order id find as a discount from table 1 and city from table 2": "SELECT df1.City,df2.Discount_Percent FROM df1 JOIN df2 ON df1.Order_Id = df2.Order_Id;",
        "13.# nested querry or sub querry give order ID Value Find Product ID": "select product_id,Selling_Price from df2 where order_id in (select order_id from df1 where order_id = '5');",
        "14.# Seasonality bases Profit": "SELECT month_name,year,day,sum(quantity) as qty from cleanedorders1 GROUP by month_name order by qty desc;",
        "15.# Segment": "SELECT segment,max(selling_price) as sales from cleanedorders1 GROUP by segment order by sales desc;",
        "16.# Discount Effectiveness": "SELECT product_id,order_date,city,category,sub_category,quantity,selling_price,max(profit) as profit, (list_price - selling_price ) / selling_price * 100 AS discount FROM cleanedorders1 WHERE selling_price GROUP by quantity ORDER BY profit>0 DES;",
        "17.# quantaty increase for discount using having method": "SELECT product_id,order_date,city,category,sub_category,quantity,selling_price,profit, (list_price - selling_price ) / selling_price * 100 AS discount FROM cleanedorders1 WHERE selling_price IS NOT NULL GROUP by quantity having quantity>0 order by quantity desc;",
        "18.# ship mode prefered most": "SELECT ship_mode,sum(quantity) as qty from cleanedorders1 GROUP by ship_mode order by qty desc;",
        "19.# Identify products with discounts greater than 20%": "SELECT product_id,selling_price,discount,discount_percent FROM cleanedorders1 WHERE discount_percent > 0.20 order by discount_percent asc;",
        "20.# Are there any seasonal trends in the data, such as higher sales during holidays or specific months?": "SELECT month_name,year,day,sum(quantity) as qty from cleanedorders1 GROUP by month_name order by qty desc;",

     },
}

# Create a dropdown to select the query set
selected_query_set_name = st.selectbox("Select Query Set:", list(query_sets.keys()))
selected_query_set = query_sets[selected_query_set_name]

# Create a dropdown to select queries within the selected set
selected_query_name = st.selectbox("Select Query:", list(selected_query_set.keys()))
selected_query = selected_query_set[selected_query_name]

# Execute the selected query based on user choice
if st.button("Execute Query"):
    # Fetch results and convert to DataFrame
    result = pd.read_sql_query(selected_query, conn)

    # Display the query results
    st.write("Query results:")
    st.dataframe(result)

# Close the connection
conn.close()


Writing streamlit5_app.py


In [101]:
!streamlit run streamlit5_app.py  &>/content/logs.txt & npx localtunnel --port 8501

⠙⠹⠸⠼⠴⠦your url is: https://tiny-toys-try.loca.lt
^C
